src/utils/staking.js

```diff
@@ -0,0 +1,278 @@
+            totalAvailable: totalAvailable.toString(),
+            totalStaked: totalStaked.toString(),
+            totalUnstaked: totalUnstaked.toString(),
+            totalUnclaimed: totalUnclaimed.toString(),
+        }
+    }
+
+    // TODO when we enable direct account staking
+    async updateStakingAccount() {
+        return {}
+    }
+
+    async getValidators() {
+
+        console.log('this.provider.validators()', (await this.provider.validators()))
+
+        return (await Promise.all(
+            (await this.provider.validators()).current_validators
+
+            // TODO remove for mainnet
+            .filter(({ account_id }) => process.env.REACT_APP_NETWORK_ID === 'betanet' ? BETANET_WHITELIST.includes(account_id) : true)
+            
+            .map(({ account_id }) => (async () => {
+                const validator = {
+                    accountId: account_id,
+                    staked: '0',
+                    unclaimed: '0',
+                    pending: '0',
+                    available: '0',
+                    contract: await this.getContractInstance(account_id, stakingMethods)
+                }
+                try {
+                    const fee = validator.fee = await validator.contract.get_reward_fee_fraction()
+                    fee.percentage = fee.numerator / fee.denominator * 100
+                    return validator
+                } catch (e) {
+                    console.warn(e)
+                }
+            })())
+        )).filter((v) => !!v)
+    }
+
+    // useLockup will be set by user in redux state and passed through actions
+    async stake(useLockup, validatorId, amount) {
+        amount = parseNearAmount(amount)
+        const { contract, lockupId } = await this.getLockup()
+        const selectedValidatorId = await contract.get_staking_pool_account_id()
+        console.log('selectedValidatorId', selectedValidatorId)
+        console.log('validatorId', validatorId)
+        if (validatorId !== selectedValidatorId) {
+            await this.lockupSelect(validatorId, lockupId, selectedValidatorId !== null)
+        }
+        return await this.lockupStake(lockupId, validatorId, amount)
+    }
+
+    async unstake(useLockup) {
+        const { lockupId } = await this.getLockup()
+        const unstake_all = await this.signAndSendTransaction(lockupId, [
+            functionCall('unstake_all', {}, GAS_STAKE)
+        ])
+        console.log('unstake_all', unstake_all)
+    }
+
+    async withdraw(useLockup) {
+        const { lockupId } = await this.getLockup()
+        const withdraw_all_from_staking_pool = await this.signAndSendTransaction(lockupId, [
+            functionCall('withdraw_all_from_staking_pool', {}, GAS_STAKE)
+        ])
+        console.log('withdraw_all_from_staking_pool', withdraw_all_from_staking_pool)
+        if (withdraw_all_from_staking_pool[0] === false) {
+            throw new Error('Unable to withdraw pending balance from validator')
```

Comment:
Should these all be `WalletError` instances so that localized message can be provided?